jbmc/src/java_bytecode/select_pointer_type.cpp

```diff
@@ -78,15 +82,28 @@ pointer_typet select_pointer_typet::specialize_generics(
       return pointer_type;
     }
     const pointer_typet &type =
-      generic_parameter_specialization_map.find(parameter_name)->second.top();
+      generic_parameter_specialization_map.find(parameter_name)->second.back();
+
+    // avoid infinite recursion
```

Comment:
Can you please move this block that handles recursion to the beginning of this `if` block? The logic of the subsequent steps would be more clear - if you're specializing a parameter, first check if you've seen it already (i.e., there's recursion), if not try to find it in the map.